yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -336,93 +349,160 @@ export class AvmEphemeralForest {
-    let start = 0;
-    let end = arr.length;
-    // Note that the easiest way is to increment the search key by 1 and then do a binary search
-    const searchKey = key.add(Fr.ONE);
-    while (start < end) {
-      const mid = Math.floor((start + end) / 2);
-      if (arr[mid].cmp(searchKey) < 0) {
-        // The key + 1 is greater than the arr element, so we can continue searching the top half
-        start = mid + 1;
-      } else {
-        // The key + 1 is LT or EQ the arr element, so we can continue searching the bottom half
-        end = mid;
-      }
-    }
-    // We either found key + 1 or start is now at the index of the largest element that we would have inserted key + 1
-    // Therefore start - 1 is the index of the element just below - note it can be -1 if the first element in the array is
-    // greater than the key
-    return start - 1;
-  }
-
   /**
-   * This gets the low leaf preimage and the index of the low leaf in the indexed tree given a value (slot or nullifier value)
-   * If the value is not found in the tree, it does an external lookup to the merkleDB
+   * Get the leaf or low leaf preimage and its index in the indexed tree given a key (slot or nullifier value).
+   * If the key is not found in the tree, it does an external lookup to the underlying merkle DB.
    * @param treeId - The tree we are looking up in
-   * @param key - The key for which we are look up the low leaf for.
+   * @param key - The key for which we are look up the leaf or low leaf for.
    * @param T - The type of the preimage (PublicData or Nullifier)
-   * @returns The low leaf preimage and the index of the low leaf in the indexed tree
+   * @returns The leaf or low leaf info (preimage & leaf index).
    */
   async getLeafOrLowLeafInfo<ID extends IndexedTreeId, T extends IndexedTreeLeafPreimage>(
     treeId: ID,
     key: Fr,
   ): Promise<PreimageWitness<T>> {
+    const [leafOrLowLeafInfo, _] = await this._getLeafOrLowLeafInfo<ID, T>(treeId, key);
+    return leafOrLowLeafInfo;
+  }
+
+  /**
+   * Internal helper to get the leaf or low leaf preimage and its index in the indexed tree given a key (slot or nullifier value).
+   * If the key is not found in the tree, it does an external lookup to the underlying merkle DB.
+   * Indicates whethe the sibling path is absent in the ephemeral tree.
+   * @param treeId - The tree we are looking up in
+   * @param key - The key for which we are look up the leaf or low leaf for.
+   * @param T - The type of the preimage (PublicData or Nullifier)
+   * @returns [
+   *     preimageWitness - The leaf or low leaf info (preimage & leaf index),
+   *     pathAbsentInEphemeralTree - whether its sibling path is absent in the ephemeral tree (useful during insertions)
+   * ]
+   */
+  async _getLeafOrLowLeafInfo<ID extends IndexedTreeId, T extends IndexedTreeLeafPreimage>(
+    treeId: ID,
+    key: Fr,
+  ): Promise<[PreimageWitness<T>, /*pathAbsentInEphemeralTree=*/ boolean]> {
```

Comment:
This never mutates the ephemeral tree and instead returns a bool that `writeStorage` and `appendNullifier` use to determine whether _they_ should update the tree with the missing path.